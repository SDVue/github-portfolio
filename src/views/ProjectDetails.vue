<template>
  <div
    v-if="details"
    class="project"
  >
    <h3 class="project__name">{{ details.name }}</h3>
    <p class="project__description">{{ details.description }}</p>
    <p class="project__url">
      <a :href="details.html_url">{{ details.html_url }}</a>
    </p>
    <router-link to="/">Home</router-link>
  </div>
</template>

<script>
import API from "@/api";
export default {
  props: {
    username: {
      type: String,
      required: true
    },
    project: {
      type: String,
      required: true
    }
  },

  data() {
    return {
      details: null
    };
  },

  async created() {
    const { username, project } = this;
    this.details = await API.getProjectDetails({ username, project });
  }
};
</script>
