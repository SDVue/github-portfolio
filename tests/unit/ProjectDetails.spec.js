import { shallowMount } from "@vue/test-utils";
import ProjectDetails from "@/views/ProjectDetails.vue";
import API from "@/api";

API.getProjectDetails = jest.fn(({ username, project }) => {
  return Promise.resolve({
    name: `Project: ${project}`,
    description: `A fancy project by ${username}.`,
    html_url: `https://github.com/${username}/${project}`
  });
});

const username = "VueSD";
const project = "github-portfolio";

describe("ProjectDetails.vue", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallowMount(ProjectDetails, {
      stubs: ["router-link"],
      propsData: {
        username,
        project
      }
    });
  });

  describe("vm.created()", () => {
    it("should call the API with props data", () => {
      expect(API.getProjectDetails).toHaveBeenCalledWith({
        username,
        project
      });
    });
    it("should update vm.details", () => {
      expect(wrapper.vm.details).toEqual({
        name: `Project: ${project}`,
        description: `A fancy project by ${username}.`,
        html_url: `https://github.com/${username}/${project}`
      });
    });
    it("should show new vm.details with new props", () => {
      const newUsername = "vuejs";
      const newProject = "vue";
      const wrapper = shallowMount(ProjectDetails, {
        stubs: ["router-link"],
        propsData: {
          username: newUsername,
          project: newProject
        }
      });
      wrapper.vm.$nextTick(() => {
        expect(API.getProjectDetails).toHaveBeenCalled();
        expect(wrapper.vm.details).toEqual({
          name: `Project: ${newProject}`,
          description: `A fancy project by ${newUsername}.`,
          html_url: `https://github.com/${newUsername}/${newProject}`
        });
      });
    });
  });

  describe("html", () => {
    it("should show project name, description, and url", () => {
      const name = wrapper.find(".project__name");
      const description = wrapper.find(".project__description");
      const url = wrapper.find(".project__url");
      const definitelyNote = wrapper.find(".definitelyNote");
      expect(name.exists()).toBe(true);
      expect(description.exists()).toBe(true);
      expect(url.exists()).toBe(true);
      expect(definitelyNote.exists()).toBe(false);
    });
    it("should include our props in the text", () => {
      let text = wrapper.text();
      expect(text.includes(username)).toBe(true);
      expect(text.includes(project)).toBe(true);
    });
  });
});
