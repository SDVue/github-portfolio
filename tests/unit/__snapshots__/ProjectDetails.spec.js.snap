// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectDetails.vue html should match the snapshot 1`] = `
<div class="project">
  <h3 class="project__name">Project: github-portfolio</h3>
  <p class="project__description">A fancy project by VueSD.</p>
  <p class="project__url"><a href="https://github.com/VueSD/github-portfolio">https://github.com/VueSD/github-portfolio</a></p>
  <router-link-stub to="/">Home</router-link-stub>
</div>
`;
